File: lib/src/main/java/me/ameriod/lib/mvp/view/MvpController.java
package me.ameriod.lib.mvp.view;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bluelinelabs.conductor.Controller;

import me.ameriod.lib.mvp.Mvp;
import me.ameriod.lib.mvp.view.deligate.ControllerDelegate;
import me.ameriod.lib.mvp.view.deligate.ControllerDelegateCallback;
import me.ameriod.lib.mvp.view.deligate.ControllerDelegateImpl;
import me.ameriod.lib.mvp.view.deligate.FragmentDelegate;

public abstract class MvpController<V extends Mvp.View, P extends Mvp.Presenter<V>> extends Controller
        implements Mvp.View, ControllerDelegateCallback<V, P> {

    protected P presenter;
    private ControllerDelegate<V, P> controllerDelegate;

    public MvpController() {
    }

    public MvpController(@Nullable Bundle args) {
        super(args);
    }

    @NonNull
    protected abstract View inflateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container);

    @NonNull
    @Override
    @SuppressWarnings("unchecked")
    protected View onCreateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
        View view = inflateView(inflater, container);
        if (controllerDelegate == null) {
            controllerDelegate = new ControllerDelegateImpl<>(this);
        }
        controllerDelegate.onCreateView();
        onPostCreateView(view, container);
        return view;
    }

    protected void onPostCreateView(@NonNull View view, @NonNull ViewGroup container) {
    }

    @Override
    protected void onRestoreViewState(@NonNull View view, @NonNull Bundle savedViewState) {
        super.onRestoreViewState(view, savedViewState);
        controllerDelegate.onRestoreViewState(savedViewState);
    }

    @Override
    protected void onSaveViewState(@NonNull View view, @NonNull Bundle outState) {
        super.onSaveViewState(view, outState);
        controllerDelegate.onSaveViewState(outState);
    }

    @Override
    protected void onDestroyView(@NonNull View view) {
        super.onDestroyView(view);
        controllerDelegate.onDestroyView();
    }

    @NonNull
    public abstract P createPresenter();

    public P getPresenter() {
        return controllerDelegate.getPresenter();
    }

    @Override
    @SuppressWarnings("unchecked")
    public V getMvpView() {
        return (V) this;
    }
}

File: lib/src/main/java/me/ameriod/lib/mvp/view/MvpController.kt
package me.ameriod.lib.mvp.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.bluelinelabs.conductor.Controller
import me.ameriod.lib.error.Error
import me.ameriod.lib.error.ErrorDisplayDelegate
import me.ameriod.lib.error.ErrorDisplayDelegateImpl

import me.ameriod.lib.mvp.Mvp
import me.ameriod.lib.mvp.deligate.ControllerDelegate
import me.ameriod.lib.mvp.deligate.ControllerDelegateCallback
import me.ameriod.lib.mvp.deligate.ControllerDelegateImpl

abstract class MvpController<V : Mvp.View, P : Mvp.Presenter<V>> : Controller, Mvp.View,
        ControllerDelegateCallback<V, P> {

    private var controllerDelegate: ControllerDelegate<V, P>? = null
    private var errorDisplayDelegate : ErrorDisplayDelegate? = null

    constructor() : super()

    constructor(args: Bundle?) : super(args)

    protected abstract fun inflateView(inflater: LayoutInflater, container: ViewGroup): View

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup): View {
        val view = inflateView(inflater, container)
        if (controllerDelegate == null) {
            controllerDelegate = ControllerDelegateImpl(this)
        }
        controllerDelegate!!.onCreateView()

        if (errorDisplayDelegate == null) {
            errorDisplayDelegate = ErrorDisplayDelegateImpl()
        }
        errorDisplayDelegate!!.attachView(view)

        onPostCreateView(view, container)

        return view
    }

    protected open fun onPostCreateView(view: View, container: ViewGroup) {}

    override fun onRestoreViewState(view: View, savedViewState: Bundle) {
        super.onRestoreViewState(view, savedViewState)
        controllerDelegate!!.onRestoreViewState(savedViewState)
    }

    override fun onSaveViewState(view: View, outState: Bundle) {
        super.onSaveViewState(view, outState)
        controllerDelegate!!.onSaveViewState(outState)
    }

    override fun onDestroyView(view: View) {
        super.onDestroyView(view)
        controllerDelegate!!.onDestroyView()
        errorDisplayDelegate!!.detachView()
    }

    abstract override fun createPresenter(): P

    override fun getPresenter(): P {
        return controllerDelegate!!.getPresenter()
    }

    @Suppress("UNCHECKED_CAST")
    override fun getMvpView(): V {
        return this as V
    }

    fun getErrorDisplayDelegate() : ErrorDisplayDelegate = errorDisplayDelegate!!

    override fun displayErrorMessage(error: Error<*>) {
        errorDisplayDelegate!!.displayError(error)
    }

    override fun displayError(error: String) {
        displayErrorMessage(Error.SnackbarMessage(error))
    }
}
